controllers/propagator/placement_change_reconclier.go

```diff
@@ -0,0 +1,190 @@
+package propagator
+
+import (
+	"context"
+	"sync"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+	appsv1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/placementrule/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	"open-cluster-management.io/governance-policy-propagator/controllers/common"
+)
+
+//+kubebuilder:rbac:groups=policy.open-cluster-management.io,resources=placementbindings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps.open-cluster-management.io,resources=placementrules,verbs=get;list;watch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PlacementChangeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	placementBindingMapper := func(ctx context.Context, obj client.Object) []reconcile.Request {
+		//nolint:forcetypeassert
+		pb := obj.(*policiesv1.PlacementBinding)
+
+		log := log.WithValues("placementBindingName", pb.GetName(), "namespace", pb.GetNamespace())
+		log.V(2).Info("Reconcile request for a PlacementBinding")
+
+		return common.GetPoliciesInPlacementBinding(ctx, r.Client, pb)
+	}
+
+	// only reconcile when the pb contains a policy or a policyset as a subject
+	pbPredicateFuncs := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			//nolint:forcetypeassert
+			pbObjNew := e.ObjectNew.(*policiesv1.PlacementBinding)
+			//nolint:forcetypeassert
+			pbObjOld := e.ObjectOld.(*policiesv1.PlacementBinding)
+
+			return common.IsForPolicyOrPolicySet(pbObjNew) || common.IsForPolicyOrPolicySet(pbObjOld)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			//nolint:forcetypeassert
+			pbObj := e.Object.(*policiesv1.PlacementBinding)
+
+			return common.IsForPolicyOrPolicySet(pbObj)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			//nolint:forcetypeassert
+			pbObj := e.Object.(*policiesv1.PlacementBinding)
+
+			return common.IsForPolicyOrPolicySet(pbObj)
+		},
+	}
+
+	// placementRuleMapper maps from a PlacementRule to the PlacementBindings associated with it
+	placementRuleMapper := func(ctx context.Context, object client.Object) []reconcile.Request {
+		log := log.WithValues("placementRuleName", object.GetName(), "namespace", object.GetNamespace())
+		log.V(2).Info("Reconcile Request for PlacementRule")
+
+		pbList := &policiesv1.PlacementBindingList{}
+		lopts := &client.ListOptions{Namespace: object.GetNamespace()}
+
+		opts := client.MatchingFields{"placementRef.name": object.GetName()}
```

Comment:
Are you able to also filter by PlacementRef.Kind == "PlacementRule"?

Reply:
I was half just following what was already in one of the mappers, and bringing it into the others. But based on this doc comment, I think it should only use one filter: https://github.com/kubernetes-sigs/controller-runtime/blob/116a1b831fffe7ccc3c8145306c3e1a3b1b14ffa/pkg/client/options.go#L422